README.md
[![Build Status](https://travis-ci.org/Fullscreen/funky.svg?branch=master)](https://travis-ci.org/Fullscreen/funky)
[![Coverage Status](https://coveralls.io/repos/github/Fullscreen/funky/badge.svg?branch=master)](https://coveralls.io/github/Fullscreen/funky?branch=master)
[![Dependency Status](https://gemnasium.com/badges/github.com/Fullscreen/funky.svg)](https://gemnasium.com/github.com/Fullscreen/funky)
[![Code Climate](http://img.shields.io/codeclimate/github/Fullscreen/funky.svg)](https://codeclimate.com/github/Fullscreen/funky)
[![Gem Version](https://badge.fury.io/rb/funky.svg)](https://badge.fury.io/rb/funky)

# Funky

Funky is a Ruby library to fetch data about videos posted on Facebook, such as their title, description, number of views, comments, shares, and likes.

## How it works

Funky can get *public* Facebook video data whether the Graph API requires insight permission or not. For example, even though the number of shares and views are shown publicly on the web page, the Graph API will not return those results unless the user has insight permission for that video. Using Funky, you can obtain the number of shares and views without insight permissions.

Under the hood, Funky hits Facebook's APIs on some cases, while other cases it will scrape Facebook's HTML to get the data. It's kind of... funky.

## Usage

This is still a very early version, and it currently can only retrieve certain Facebook video data.

### Installing and Configuring Funky

First, add funky to your Gemfile:

```ruby
gem 'funky'
```
Then run `bundle install`.

Funky will require an App ID and an App Secret which you can obtain after registering as a developer on [Facebook for developers](https://developers.facebook.com/).

There are two ways to configure Funky with your App ID and App Secret:

1. By default, Funky will look for the environment variables called `FB_APP_ID` and `FB_APP_SECRET`. You can put those keys in your `.bash_profile` and Funky will work.

    ```
    export FB_APP_ID="YourAppID"
    export FB_APP_SECRET="YourAppSecret"
    ```

2. If you're using this in Rails, you can choose to create an initializer instead and configure the App ID and App Secret like so:

    ```ruby
    Funky.configure do |config|
      config.app_id = 'YourAppID'
      config.app_secret = 'YourAppSecret'
    end
    ```

## API Overview

Funky consists of 2 different surface APIs - one to fetch video data
from Facebook and one to fetch page data from Facebook.

## Pages API

### Use #where clause to get an array of pages

```ruby
ids = ['1487249874853741', '526533744142224']
pages = Funky::Page.where(id: ids)
pages.first.id            # => '1487249874853741'
pages.first.name          # => 'Sony Pictures'
pages.first.username      # => 'SonyPicturesGlobal'
```

### Use #posts to get an array of posts

```ruby
page = Funky::Page.find('FullscreenInc')
posts = page.posts
posts.first.type          # => 'video'
```

## Videos API

### Use #where clause to get an array of videos

```ruby
ids = ['10154439119663508', '10153834590672139']
videos = Funky::Video.where(id: ids)
videos.first.id            # => '10154439119663508'
videos.first.created_time  # => #<DateTime: 2015-12-17T06:29:48+00:00>
videos.first.description   # => "Hugh Jackman coaches Great Britain's..."
videos.first.length        # => 147.05
videos.first.picture       # => "https://scontent.xx.fbcdn.net/v/..."
videos.first.page_name     # => "Moviefone"

```

If a non-existing video ID is passed into the where clause, it is ignored. Other video IDs will still be retrieved.

```ruby
ids = ['10154439119663508', '10153834590672139', 'doesnotexist']
videos = Funky::Video.where(id: ids)
videos.count    # => 2
videos.first.id # => '10154439119663508'
videos.last.id  # => '10153834590672139'
```

### Use #find to get a single video

```ruby
video = Funky::Video.find('10154439119663508')
video.id            # => '10154439119663508'
video.like_count    # => 1169
video.comment_count # => 65
video.share_count   # => 348
video.view_count    # => 10121
```

If a non-existing video ID is passed into #find, Funky::ContentNotFound will be raised.

```ruby
Funky::Video.find('doesnotexist') # => raises Funky::ContentNotFound
```

### Use #find_by_url!(url) to get a single video from a url

```ruby
video = Funky::Video.find_by_url!('https://www.facebook.com/video.php?v=10154439119663508')
video.id            # => '10154439119663508'
video.like_count    # => 1169
video.comment_count # => 65
video.share_count   # => 348
video.view_count    # => 10121
```

If a non-existing video url is passed into #find_by_url!, Funky::ContentNotFound will be raised.

```ruby
Funky::Video.find_by_url!('https://www.facebook.com/video.php?v=doesnotexist')
  # => raises Funky::ContentNotFound
```

The current URL formats are supported:

- `https://www.facebook.com/{page_name}/videos/vb.{alt_page_id}/{video_id}`
- `https://www.facebook.com/{page_name}/videos/{video_id}`
- `https://www.facebook.com/{page_id}/videos/{video_id}`
- `https://www.facebook.com/video.php?v={video_id}`


### Connection error

Should there be a case where Funky is unable to connect to Facebook, `Funky::ConnectionError` will be raised.

```ruby
# Given funky is unable to establish a connection to Facebook
Funky::Video.find('10154439119663508') # => raises Funky::ConnectionError
Funky::Video.where(id: '10154439119663508') # => raises Funky::ConnectionError
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/fullscreen/funky. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.


## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

For more information about changelogs, check
[Keep a Changelog](http://keepachangelog.com) and
[Vandamme](http://tech-angels.github.io/vandamme).


## 0.2.38 - 2019/10/07

* [BUGFIX] Fetch like count for more Facebook videos.

## 0.2.37 - 2019/07/23 (Do not use)

* [BUGFIX] Allow funky to scrape view count for more videos

## 0.2.36 - 2019/03/05

* [BUGFIX] Fetch comment count, share count, like count for Facebook
videos with different HTML.

## 0.2.35 - 2019/01/14

* [ENHANCEMENT] Use 2.10 of Facebook Graph API (upgrade from 2.8).
  - There will be deprecations in v2.8 of the Facebook Graph API soon. Check out [the changelog](https://developers.facebook.com/docs/apps/changelog).
  - Funky API is not changed. (The same test is still passing)

## 0.2.34 - 2018/10/19

* [BUGFIX] Fetch the comment count, with more Facebook videos.

## 0.2.33 - 2018/10/03

* [FEATURE] Add ActiveSupport::Notification based instrumentation.

## 0.2.32 - 2018/06/07

* [FEATURE] Raise Funky::CountersNotFound error when a video does not have
its view count.

## 0.2.31 - 2017/10/25

* [BUGFIX] Revert the change of version 0.2.29 and do not use request to get app
access token.
* [ENHANCEMENT] Allow funky to scrape view count for more videos, by changing
the way how to scrape.

## 0.2.30 - 2017/10/16

* [BUGFIX] Test should pass when call Page#videos. Partly revert the change
of version 0.2.28

## 0.2.29 - 2017/10/16

* [ENHANCEMENT] Use request to get app access token. Do not use
`app_id|app_secret` string.

## 0.2.28 - 2017/10/05

* [FEATURE] Add `since` option when it calls Page#videos and Page#posts to reduce
amount of requests when it gets only recent items.

## 0.2.27 - 2017/08/28

* [BUGFIX] Fetch the correct view count even for non-English pages

## 0.2.26 - 2017/06/06

* [ENHANCEMENT] Raise ContentNotFound if Page#find is called with an invalid URI

## 0.2.25 - 2017/05/16

* [FEATURE] Retry up to 5 times if Facebook responds with error

## 0.2.24 - 2017/05/10

* [BUGFIX] Raise Funky::ContentNotFound error when a page does not return
its name. This change will filter out cases a full url of website entered
as page_id argument of `Funky::Page#find`, instead of username.

## 0.2.23 - 2017/05/09

* [FEATURE] Print out logs for each pagination
* [BUGFIX] Use cursor-based pagination for `Funky::Page#posts` to resolve
infinite loop issue when the page has life events with old backdated time.

## 0.2.22 - 2017/05/05

* [FEATURE] Add `Funky::Page#posts`.
* [FEATURE] Add `Funky::Page#has_featured_video?`.

## 0.2.21 - 2017/05/03

* [FEATURE] Add `Funky::Page#fan_count`.

## 0.2.20 - 2017/04/20

* [BUGFIX] Fix `Funky::Page#videos` by adding condition for an edge case.

## 0.2.19 - 2017/04/20

* [IMPROVEMENT] Retry 3 times after a server error or a socket error, to fetch
  data even with any temporary issue from Facebook.

## 0.2.18 - 2017/04/19

* [ENHANCEMENT] `Funky::Page#videos` now fetches more videos by time-based pagination.

## 0.2.17 - 2017/04/17

* [FEATURE] Add `Funky::Page.find` class method to find a page by its page ID.
* [FEATURE] Add `Funky::Page#videos` instance method to list all videos under a page fetched by API call.

## 0.2.16 - 2017/04/02

* [ENHANCEMENT] Video.where(id: [..ids..]) now works even with more than 50 ids.

## 0.2.15 - 2017/03/21

* [BUGFIX] Return a string for the description field of a Funky::Video
  object even when the Facebook video has no description

## 0.2.14 - 2017/02/06

* [ENHANCEMENT] Add the Funky::Page API to fetch name, username,
  location, city, state, zip, street, country, longitude,
  and latitude fields from the Facebook Graph API given a page ID.

## 0.2.13 - 2017/01/25

* [BUGFIX] Correctly fetch data for videos that belong to a Facebook page with a username that contains URL-unsafe characters. For instance "https://www.facebook.com/KinoToPrzygoda /" (with a space) is a valid Facebook page URL.
* [ENHANCEMENT] Use 2.8 of Facebook Graph API (upgrade from 2.6).
- There were deprecations in v2.8 of the Facebook Graph API as well as additions documented in [the changelog](https://developers.facebook.com/docs/apps/changelog).
- The current API of Funky is not affected.

## 0.2.12 - 2017/01/19

* [ENHANCEMENT] Parse new format of share & comment count in Facebook HTML
* [BUGFIX] Fix regular expression for likecount

## 0.2.11 - 2017/01/19

* [ENHANCEMENT] Follow redirects when Facebook responds with 302
* [ENHANCEMENT] Parse new format of view count in Facebook HTML

## 0.2.10 - 2016/12/20

* [ENHANCEMENT] Add a facebook page ID as a part of the Funky::Video API
* [ENHANCEMENT] Add a facebook page URL as a part of the Funky::Video API

## 0.2.9 - 2016/12/13

* [BUGFIX] Allow a video ID to be parsed out of facebook video URLs with
trailing slashes.

## 0.2.8 - 2016/12/12

* [BUGFIX] Change the way that "like counts" are parsed now that Facebook has changed the structure of the HTML and can include multiple "likecounts" elements in the same page.

## 0.2.7 - 2016/08/18

* [ENHANCEMENT] When Funky makes a request to an HTML page, Funky would now raise `Funky::ConnectionError` should the following errors occur:
OpenSSL::SSL::SSLError,
Errno::ETIMEDOUT,
Errno::EHOSTUNREACH,
Errno::ENETUNREACH,
Errno::ECONNRESET,
Net::OpenTimeout,
SocketError

## 0.2.6 - 2016/08/05

* [BUGFIX] Force scraping to use English-US version of Facebook by appending `locale=en_US` to the query string. This prevents cases where Funky's requests originate from non English speaking countries and receiving different HTML templates.

## 0.2.5 - 2016/07/25

* [ENHANCEMENT] Allow Funky to also scrape view count that uses dot delimiters. Previously, it only scraped view count with comma delimiters.

## 0.2.4 - 2016/06/28


lib/funky/video.rb
require 'funky/html/page'
require 'funky/html/parser'
require 'funky/url'

module Funky
  class Video < GraphRootNode

    attr_accessor :counters

    @@html_page = HTML::Page.new
    @@html_parser = HTML::Parser.new

    # @return [DateTime] the created time of the video.
    def created_time
      datetime = data[:created_time]
      DateTime.parse datetime if datetime
    end

    # @return [String] the description of the video.
    def description
      data[:description].to_s
    end

    # @return [Float] the length (duration) of the video.
    def length
      data[:length]
    end

    # @return [String] the title of the video.
    def title
      data[:title].to_s
    end

    # see https://developers.facebook.com/docs/graph-api/reference/video/
    # @return [Integer] the total number of likes for the video.
    def count_likes
      data[:likes][:summary][:total_count]
    end

    # @return [Integer] the total number of comments for the video.
    def count_comments
      data[:comments][:summary][:total_count]
    end

    # @return [Integer] the total number of reactions for the video.
    def count_reactions
      data[:reactions][:summary][:total_count]
    end

    # @return [String] the picture URL of the video.
    def picture
      data[:picture]
    end

    # @return [String] the name of Facebook page for the video.
    def page_name
      data.fetch(:from)[:name]
    end

    # @return [String] the id of Facebook page for the video.
    def page_id
      data.fetch(:from)[:id]
    end

    # @return [String] the url of Facebook page for the video.
    def page_url
      "https://www.facebook.com/#{page_id}"
    end

    # @return [Integer] the total number of likes for the video.
    def like_count
      data[:like_count]
    end

    # @return [Integer] the total number of comments for the video.
    def comment_count
      data[:comment_count]
    end

    # @return [Integer] the total number of shares for the video.
    def share_count
      data[:share_count]
    end

    # @return [Integer] the total number of views for the video.
    def view_count
      data[:view_count]
    end

    # Fetches the data from Facebook's APIs and instantiates the data
    # into an Array of Funky::Video objects. It can accept one video ID or an
    # array of multiple video IDs.
    #
    # @example Getting one video
    #   id = '10153834590672139'
    #   Funky::Video.where(id: id) # => [#<Funky::Video>]
    # @example Getting multiple videos
    #   ids = ['10154439119663508', '10153834590672139']
    #   Funky::Video.where(id: ids) # => [#<Funky::Video>, #<Funky::Video>]
    #
    # @return [Array<Funky::Video>] multiple instances of Funky::Video objects
    #   containing data obtained by Facebook's APIs.
    def self.where(id:)
      return nil unless id
      instantiate_collection(fetch_and_parse_data Array(id))
    end

    # Fetches the data from Facebook's HTML and instantiates the data
    # into a single Funky::Video object. It can accept one only video ID.
    #
    # @example Getting a video
    #   Funky::Video.find('10153834590672139') # => #<Funky::Video>
    #
    # @return [Funky::Video] the data scraped from Facebook's HTML
    #   and encapsulated into a Funky::Video object.
    def self.find(video_id)
      counters = @@html_parser.parse html: @@html_page.get(video_id: video_id), video_id: video_id
      # raise CountersNotFound, "View count not found with video ID #{video_id}" unless counters[:view_count]
      new counters.merge(id: video_id)
    end

    # Similar to #find, but it finds the video by url instead of video id.
    # Fetches the data from Facebook's HTML and instantiates the data
    # into a single Funky::Video object. It can accept one only video url.
    #
    # @example Getting a video by url
    #   url = 'https://www.facebook.com/video.php?v=203203106739575'
    #   Funky::Video.find_by_url!(url) # => #<Funky::Video>
    #
    # @return [Funky::Video] the data scraped from Facebook's HTML
    #   and encapsulated into a Funky::Video object.
    def self.find_by_url!(url)
      url = URL.new(url)
      find(url.video_id)
    end

  private

    def self.fields
      [
        'created_time',
        'description',
        'length',
        'from',
        'picture'
      ]
    end
  end
end
